reference/wbOptions.md
# WinBox Options

WinBox Options

## Usage

``` r
wbOptions(
  width = NULL,
  height = NULL,
  minwidth = NULL,
  minheight = NULL,
  x = NULL,
  y = NULL,
  max = NULL,
  min = NULL,
  top = NULL,
  right = NULL,
  bottom = NULL,
  left = NULL,
  background = NULL,
  border = NULL,
  modal = NULL,
  index = NULL,
  ...
)
```

## Arguments

- width, height:

  Set the initial width/height of the window (supports units "px" and
  "%").

- minwidth, minheight:

  Set the minimal width/height of the window (supports units "px" and
  "%").

- x, y:

  Set the initial position of the window (supports: "right" for x-axis,
  "bottom" for y-axis, "center" for both, units "px" and "%" for both).

- max, min:

  Automatically toggles the window into maximized / minimized state when
  created.

- top, right, bottom, left:

  Set or limit the viewport of the window's available area (supports
  units "px" and "%").

- background:

  Set the background of the window (supports all CSS styles which are
  also supported by the style-attribute "background", e.g. colors,
  transparent colors, hsl, gradients, background images).

- border:

  Set the border width of the window (supports all css units, like px,
  %, em, rem, vh, vmax).

- modal:

  Shows the window as modal.

- index:

  Set the initial z-index of the window to this value (could be
  increased automatically when unfocused/focused).

- ...:

  Other options, see
  https://github.com/nextapps-de/winbox?tab=readme-ov-file#options.

## Value

A `list` of options to use in
[`WinBox()`](https://dreamrs.github.io/shinyWidgets/reference/WinBox.md).

## Examples

``` r

library(shiny)
library(shinyWidgets)

ui <- fluidPage(
  html_dependency_winbox(),
  actionButton(inputId = "show1", label = "Show WinBox"),
  actionButton(inputId = "show2", label = "Show WinBox as modal")
)

server <- function(input, output, session) {

  observeEvent(input$show1, {
    WinBox(
      title = "Custom background color and border",
      ui = tagList(
        tags$h2("Hello from WinBox!"),
        "Text content of winbox."
      ),
      options = wbOptions(
        background = "#112446",
        border = "0.5em",
        x = "center",
        y = "center",
        width = "50%",
        height = "50%"
      )
    )
  })

  observeEvent(input$show2, {
    WinBox(
      title = "WinBox as modal",
      ui = tagList(
        tags$h2("Hello from WinBox!"),
        "Text content of winbox."
      ),
      options = wbOptions(modal = TRUE)
    )
  })

}

if (interactive())
  shinyApp(ui, server)
```


reference/pickerOptions.md
# Options for `pickerInput`

Wrapper of options available here:
https://developer.snapappointments.com/bootstrap-select/options/

## Usage

``` r
pickerOptions(
  actionsBox = NULL,
  container = NULL,
  countSelectedText = NULL,
  deselectAllText = NULL,
  dropdownAlignRight = NULL,
  dropupAuto = NULL,
  header = NULL,
  hideDisabled = NULL,
  iconBase = NULL,
  liveSearch = NULL,
  liveSearchNormalize = NULL,
  liveSearchPlaceholder = NULL,
  liveSearchStyle = NULL,
  maxOptions = NULL,
  maxOptionsText = NULL,
  mobile = NULL,
  multipleSeparator = NULL,
  noneSelectedText = NULL,
  noneResultsText = NULL,
  selectAllText = NULL,
  selectedTextFormat = NULL,
  selectOnTab = NULL,
  showContent = NULL,
  showIcon = NULL,
  showSubtext = NULL,
  showTick = NULL,
  size = NULL,
  style = NULL,
  tickIcon = NULL,
  title = NULL,
  virtualScroll = NULL,
  width = NULL,
  windowPadding = NULL,
  ...
)
```

## Arguments

- actionsBox:

  When set to true, adds two buttons to the top of the dropdown menu
  (Select All & Deselect All). Type: boolean; Default: false.

- container:

  When set to a string, appends the select to a specific element or
  selector, e.g., container: 'body' \| '.main-body' Type: string \|
  false; Default: false.

- countSelectedText:

  Sets the format for the text displayed when selectedTextFormat is
  `count` or `count > #`. `{0}` is the selected amount. `{1}` is total
  available for selection. When set to a function, the first parameter
  is the number of selected options, and the second is the total number
  of options. The function must return a string. Type: string \|
  function; Default: function.

- deselectAllText:

  The text on the button that deselects all options when actionsBox is
  enabled. Type: string; Default: 'Deselect All'.

- dropdownAlignRight:

  Align the menu to the right instead of the left. If set to 'auto', the
  menu will automatically align right if there isn't room for the menu's
  full width when aligned to the left. Type: boolean \| 'auto'; Default:
  false.

- dropupAuto:

  checks to see which has more room, above or below. If the dropup has
  enough room to fully open normally, but there is more room above, the
  dropup still opens normally. Otherwise, it becomes a dropup. If
  dropupAuto is set to false, dropups must be called manually. Type:
  boolean; Default: true.

- header:

  adds a header to the top of the menu; includes a close button by
  default Type: string; Default: false.

- hideDisabled:

  removes disabled options and optgroups from the menu
  data-hide-disabled: true Type: boolean; Default: false.

- iconBase:

  Set the base to use a different icon font instead of Glyphicons. If
  changing iconBase, you might also want to change tickIcon, in case the
  new icon font uses a different naming scheme. Type: string; Default:
  'glyphicon'.

- liveSearch:

  When set to true, adds a search box to the top of the selectpicker
  dropdown. Type: boolean; Default: false.

- liveSearchNormalize:

  Setting liveSearchNormalize to true allows for accent-insensitive
  searching. Type: boolean; Default: false.

- liveSearchPlaceholder:

  When set to a string, a placeholder attribute equal to the string will
  be added to the liveSearch input. Type: string; Default: null.

- liveSearchStyle:

  When set to 'contains', searching will reveal options that contain the
  searched text. For example, searching for pl with return both Apple,
  Plum, and Plantain. When set to 'startsWith', searching for pl will
  return only Plum and Plantain. Type: string; Default: 'contains'.

- maxOptions:

  When set to an integer and in a multi-select, the number of selected
  options cannot exceed the given value. This option can also exist as a
  data-attribute for an `<optgroup>`, in which case it only applies to
  that `<optgroup>`. Type: integer \| false; Default: false.

- maxOptionsText:

  The text that is displayed when maxOptions is enabled and the maximum
  number of options for the given scenario have been selected. If a
  function is used, it must return an array. `array[0]` is the text used
  when maxOptions is applied to the entire select element. `array[1]` is
  the text used when maxOptions is used on an optgroup. If a string is
  used, the same text is used for both the element and the optgroup.
  Type: string \| array \| function; Default: function.

- mobile:

  When set to true, enables the device's native menu for select menus.
  Type: boolean; Default: false.

- multipleSeparator:

  Set the character displayed in the button that separates selected
  options. Type: string; Default: ', '.

- noneSelectedText:

  The text that is displayed when a multiple select has no selected
  options. Type: string; Default: 'Nothing selected'.

- noneResultsText:

  The text displayed when a search doesn't return any results. Type:
  string; Default: `'No results matched {0}'`.

- selectAllText:

  The text on the button that selects all options when actionsBox is
  enabled. Type: string; Default: 'Select All'.

- selectedTextFormat:

  Specifies how the selection is displayed with a multiple select.
  'values' displays a list of the selected options (separated by
  multipleSeparator. 'static' simply displays the select element's
  title. 'count' displays the total number of selected options. 'count
  \> x' behaves like 'values' until the number of selected options is
  greater than x; after that, it behaves like 'count'. Type: 'values' \|
  'static' \| 'count' \| 'count \> x' (where x is an integer); Default:
  'values'.

- selectOnTab:

  When set to true, treats the tab character like the enter or space
  characters within the selectpicker dropdown. Type: boolean; Default:
  false.

- showContent:

  When set to true, display custom HTML associated with selected
  option(s) in the button. When set to false, the option value will be
  displayed instead. Type: boolean; Default: true.

- showIcon:

  When set to true, display icon(s) associated with selected option(s)
  in the button. Type: boolean; Default: true.

- showSubtext:

  When set to true, display subtext associated with a selected option in
  the button. Type: boolean; Default: false.

- showTick:

  Show checkmark on selected option (for items without multiple
  attribute). Type: boolean; Default: false.

- size:

  When set to 'auto', the menu always opens up to show as many items as
  the window will allow without being cut off. When set to an integer,
  the menu will show the given number of items, even if the dropdown is
  cut off. When set to false, the menu will always show all items. Type:
  'auto' \| integer \| false; Default: 'auto'.

- style:

  When set to a string, add the value to the button's style. Type:
  string \| null; Default: null.

- tickIcon:

  Set which icon to use to display as the "tick" next to selected
  options. Type: string; Default: 'glyphicon-ok'.

- title:

  The default title for the selectpicker. Type: string \| null; Default:
  null.

- virtualScroll:

  If enabled, the items in the dropdown will be rendered using
  virtualization (i.e. only the items that are within the viewport will
  be rendered). This drastically improves performance for selects with a
  large number of options. Set to an integer to only use virtualization
  if the select has at least that number of options. Type: boolean \|
  integer; Default: 600.

- width:

  When set to auto, the width of the selectpicker is automatically
  adjusted to accommodate the widest option. When set to a css-width,
  the width of the selectpicker is forced inline to the given value.
  When set to false, all width information is removed. Type: 'auto' \|
  'fit' \| css-width \| false (where css-width is a CSS width with
  units, e.g. 100px); Default: false.

- windowPadding:

  This is useful in cases where the window has areas that the dropdown
  menu should not cover - for instance a fixed header. When set to an
  integer, the same padding will be added to all sides. Alternatively,
  an array of integers can be used in the format
  `top, right, bottom, left`. Type: integer \| array; Default: 0.

- ...:

  Other options not listed here.

## Note

Documentation is from [Bootstrap-select onlin
page](https://developer.snapappointments.com/bootstrap-select/options/).

## Examples

``` r

if (interactive()) {
  library(shiny)
  library(shinyWidgets)

  ui <- fluidPage(
    pickerInput(
      inputId = "month",
      label = "Select a month",
      choices = month.name,
      multiple = TRUE,
      options = pickerOptions(
        actionsBox = TRUE,
        title = "Please select a month",
        header = "This is a title"
      )
    )
  )

  server <- function(input, output, session) {

  }

  shinyApp(ui, server)
}
```


NEWS.md
shinyWidgets 0.9.2
======================

* Fixed an issue in airDatepickerInput() where update_on = "close" did not work as expected [#766](https://github.com/dreamRs/shinyWidgets/issues/766).


shinyWidgets 0.9.1
======================

* New widget `quercusInput()`, hierarchical treeview component based on JavaScript library [quercus.js](https://github.com/stefaneichert/quercus.js).
* Updated virtual-select to 1.1.0, see https://github.com/sa-si-dev/virtual-select/releases.
* Updated air-datepicker to 3.6.0, see https://github.com/t1m0n/air-datepicker?tab=readme-ov-file#v360
* Updated vanilla-calendar-pro to 3.1.0, see https://github.com/uvarov-frontend/vanilla-calendar-pro/releases


shinyWidgets 0.9.0
======================

* New widget `calendarProInput()`, a date and time picker component based on JavaScript library [vanilla-calendar-pro](https://github.com/uvarov-frontend/vanilla-calendar-pro).
* `virtualSelectInput()`: ability to set a JavaScript function as callback for `selectedLabelRenderer`.
* New argument `inputType` in `searchInput()` and `textInputIcon()` to specify the input's type, for example `"password"`.
* `knobInput()` and `switchInput()` are softly deprecated, since the JavaScript libraries used by those widgets are no longer actively maintained.
* Updated virtual-select to 1.0.47 (fix focus issue), see https://github.com/sa-si-dev/virtual-select/releases.
* Updated @simonwep/pickr to 1.9.1, see https://github.com/simonwep/pickr/releases.


shinyWidgets 0.8.7
======================

* New widget : `slimSelectInput()`, an alternative select menu.
* Updated virtual-select to 1.0.45, see https://github.com/sa-si-dev/virtual-select/releases.
* `airDatepickerInput()` & `updateAirDateInput()`: added a `tz` argument to specify timezone.
* `checkboxGroupButtons()` & `radioGroupButtons()`: fixed justified layout when using individual buttons.


shinyWidgets 0.8.6
======================

* Fixed a bug in `updateAirDateInput()` when updating selected value with no options.
* Updated virtual-select to 1.0.43, see https://github.com/sa-si-dev/virtual-select/releases.


shinyWidgets 0.8.5
======================

* `WinBox()`: added `auto_index` argument to automatically open window above all others.
* Added `applyWinBox()` to apply arbitrary method on an existing window.


shinyWidgets 0.8.4
======================

* Fixed a bug in `updateAirDateInput()` when updating selected value and options at the same time [#684](https://github.com/dreamRs/shinyWidgets/issues/684).
* New widget `WinBox()` to manage multiple windows in an app, https://nextapps-de.github.io/winbox/.


shinyWidgets 0.8.3
======================

* Updated virtual-select-plugin to 1.0.42 and inclued tooltip plugin, [#674](https://github.com/dreamRs/shinyWidgets/pull/674) by [@stla](https://github.com/stla).
* Updated air-datepicker to 3.5.0.
* Updated noUiSlider to 15.7.1.
* `updateNoUiSliderInput()`: added `disableHandlers` and `enableHandlers` to disable/enable specific handlers.
* `updateVirtualSelect()` : added `open` argument to open/close the dropdown.
* `virtualSelectInput()` added `updateOn` argument to to set when input is updated: on change or on close.
* New widget `timeInput()` to select time using browser input.


shinyWidgets 0.8.2
======================

* New feature for `virtualSelectInput()` : ability to use JavaScript functions for `onServerSearch` or `labelRenderer` parameters, by [@MichalLauer](https://github.com/MichalLauer).
* New feature for `virtualSelectInput()` / `prepare_choices()` : added classNames as optional argument, by [@SverreFL](https://github.com/SverreFL).
* Updated air-datepicker to 3.4.0 (with croatian and bulgarian locale), fix [#668](https://github.com/dreamRs/shinyWidgets/issues/668).
* `updatePickerInput()` : fix max option group, fix [#670](https://github.com/dreamRs/shinyWidgets/issues/670).
* Fix JavaScript errors generated by `updateProgressBar()`, fix [#656](https://github.com/dreamRs/shinyWidgets/issues/656).


shinyWidgets 0.8.1
======================

* `searchInput()` : added `btnClass` argument to set class of search and reset buttons.
* Addition of autocomplete parameter for `pickerInput()` and `multiInput()`, [#652](https://github.com/dreamRs/shinyWidgets/pull/652) by [@MichalLauer](https://github.com/MichalLauer).

### Bug fixes

* `pickerInput()` : fix choicesOpts with grouped choices [#650](https://github.com/dreamRs/shinyWidgets/issues/650)
* `updateProgressBar()` : accept HTML in title [#651](https://github.com/dreamRs/shinyWidgets/issues/651)
* don't use system.file for packer htmldependencies
* `searchInput()` : remove inline CSS and `!important` use [#637](https://github.com/dreamRs/shinyWidgets/issues/637)


shinyWidgets 0.8.0
======================

* `shinyWidgetsGallery()` new look for the gallery and two new tabs (for `airDatepickerInput()` and `virtualSelectInput()`).
* `airDatepickerInput()` : added arguments :
    - `disabledDaysOfWeek` to disable day(s) of the week.
    - `readonly` to prevent edit in the input field.
    - `onkeydown` to add onkeydown attribute on the input field.
* `materialSwitch()` : click on the label now toggle the switch [#582](https://github.com/dreamRs/shinyWidgets/issues/582).
* `pickerInput()` : added argument `stateInput` to activate or deactivate the special input value `input$<inputId>_open` allowing to know if the menu is opened or not.
* `updateColorPickr()` : added argument `swatches` to update swatches from server.
* Updated virtual-select-plugin to 1.0.39 (fix [#601](https://github.com/dreamRs/shinyWidgets/issues/601)).
* Updated sweetalert2 to v11.7.27.
* Updated air-datepicker to 3.3.5.


### Bug fixes

* `verticalTabsetPanel()` does not render output content when using Bootstrap 5 [#570](https://github.com/dreamRs/shinyWidgets/issues/570).
* `virtualSelectInput` : fixed crash with choices=list() [#571](https://github.com/dreamRs/shinyWidgets/issues/571).


### Deprecated functions


inst/examples/WinBox/options.R

library(shiny)
library(shinyWidgets)

ui <- fluidPage(
  html_dependency_winbox(),
  actionButton(inputId = "show1", label = "Show WinBox"),
  actionButton(inputId = "show2", label = "Show WinBox as modal")
)

server <- function(input, output, session) {

  observeEvent(input$show1, {
    WinBox(
      title = "Custom background color and border",
      ui = tagList(
        tags$h2("Hello from WinBox!"),
        "Text content of winbox."
      ),
      options = wbOptions(
        background = "#112446",
        border = "0.5em",
        x = "center",
        y = "center",
        width = "50%",
        height = "50%"
      )
    )
  })

  observeEvent(input$show2, {
    WinBox(
      title = "WinBox as modal",
      ui = tagList(
        tags$h2("Hello from WinBox!"),
        "Text content of winbox."
      ),
      options = wbOptions(modal = TRUE)
    )
  })

}

if (interactive())
  shinyApp(ui, server)
